### PR TITLE
feat(math): add exact hard-constraint rounding

### LINKED ISSUE #2318: [Gap]: Exact rounding with hard cardinality constraints and column-sparse error bounds
## Observed mathematical task(s)

A finite floating-variable argument in the Erdős 390 proof starts from a rational vector `x in [0,1]^A`, a partition of the coordinates into disjoint rows whose sums are integers, and a family of monitored 0/1 columns. It repeatedly moves in an exact nullspace until another coordinate reaches `0` or `1`.

The resulting binary vector preserves every row sum exactly while changing each monitored column sum by at most `4d`, where `d` is the maximum number of monitored columns containing one coordinate.

Source evidence:

- https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/390/lean/Erdos390/WholePaper/FloatingRoundingStep.lean
- https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/390/paper.tex#L9780-L9945
- Benjamin Doerr, “Roundings Respecting Hard Constraints,” *Theory of Computing Systems* 40 (2007), 467–483: https://doi.org/10.1007/s00224-006-1318-7

This is the hard-cardinality-constraint rounding shape, not ordinary independent randomized rounding and not merely minimum discrepancy over unrestricted `+-1` colorings.

## Missing mathematical result

Given:

- a finite indexed coordinate set;
- exact rational values `0 <= x_a <= 1`;
- a partition into disjoint row constraints with integral source row sums; and
- a finite indexed family of monitored subsets/0-1 columns;

return a binary vector `X` satisfying

```text
sum_(a in row) X_a = sum_(a in row) x_a
```

for every row, together with a uniform column-error guarantee

```text
|sum_(a in column) (X_a - x_a)| <= 4d,
```

where `d` is derived from the maximum monitored-column incidence of a coordinate.

The result should retain the source data and expose complete row and column ledgers so both preservation and error bounds replay exactly.

## Bespoke escape used

The proof implements the floating-variable algorithm directly:

1. take the current fractional support `F`;
2. retain all active row equations;
3. retain monitored columns with more than `4d` active coordinates;
4. use the incidence count to prove that the retained system has fewer equations than `|F|`;
5. compute a nonzero exact nullspace direction;
6. move to an endpoint of the feasible interval so at least one more coordinate becomes integral; and
7. repeat at most once per coordinate.

The local Lean file formalizes the rank-deficient step. The paper assembles the iteration and the final discrepancy bound.

## Gap classification

- [ ] representation
- [ ] interoperability
- [ ] discovery
- [ ] contract
- [ ] scale/backend
- [x] operation
- [ ] reasoning

## Admission posture (operation gaps only)

This supports public-catalog consideration as one bounded constructive rounding operation. The reusable postcondition is exact preservation of disjoint cardinality constraints plus a source-derived column-sparse error bound. The theorem-specific allocation and later analytic argument remain caller workflow.

A plausible operation ID is:

```text
combinatorics.set_system.hard_constraint_round.compute
```

The name and domain owner should be settled during admission; the mathematical contract matters more than mirroring the source terminology.

## Existing math.find queries

Relevant searches include:

```text
finite set-system discrepancy
round fractional vector preserving row sums
column-sparse discrepancy
exact rational nullspace rounding
```

The closest current catalog surface is finite set-system discrepancy evaluation/optimization from #1693 plus exact rational linear algebra.

## Existing operations inspected

- The discrepancy family in #1693 evaluates or optimizes unrestricted `+-1` colorings. It does not accept a fractional target, preserve exact row sums, or return the `4d` hard-constraint guarantee.
- Exact rational nullspace and rational-linear operations can supply one algebraic substep, but the caller would still have to select retained equations, find a boundary step, iterate, and prove the global preservation/error postcondition. Their source-binding and scale repairs are tracked in #2310 and #2302.
- LP operations can encode feasibility for one finite instance, but a feasible binary point does not establish this constructive rounding guarantee or its source-derived bound.

## Why composition was insufficient

The indispensable result is not merely a nullspace vector or a low-discrepancy coloring. It is one binary rounding whose exact row constraints and uniform monitored-column errors are jointly guaranteed from the accepted source.

Treating the iterative support reduction as caller-side glue would require every agent to reimplement the mathematical algorithm and its global invariant. Conversely, a general mixed-integer optimizer would broaden the contract and lose the theorem-backed `4d` semantics.

## Evidence and downstream uses

Defining invariants and fixtures should include:

- every output coordinate is exactly `0` or `1`;
- every partition row reconstructs its exact source sum;
- every monitored-column source sum, rounded sum, signed error, and absolute error is returned and replayed;
- the derived incidence parameter `d` and `4d` bound agree with the source incidence relation;
- integral input is returned unchanged under the deterministic convention;
- singleton rows, empty monitored families, duplicate indexed columns, `d=0`, and coordinates appearing in no monitored column;
- rejection of non-partitions, out-of-range rationals, nonintegral row sums, missing/extra coordinate keys, and malformed incidence;
- invariance under coherent row, column, and coordinate relabeling;
- authored-result mutations to one bit, one row ledger, one column error, and the source object;
- exact requests at and immediately above coordinate, row, column, incidence, rational-digit, iteration, intermediate-height, and output bounds.

The implementation can use maintained exact rational/integer linear algebra for nullspaces, but admission must bound intermediate coefficient growth as well as dimensions. Repeated rational endpoint moves make a dimension-only cap insufficient.

Non-goals:

- minimum-discrepancy optimization;
- arbitrary overlapping hard constraints;
- preservation of probabilistic marginals or negative correlation;
- a generic rounding callback/workflow;
- theorem-specific Erdős 390 allocation; or
- asymptotic discrepancy conclusions.

## Related

- #1693 — finite set-system discrepancy evaluation and optimization.
- #2302 — sparse/rectangular rational-linear scale.
- #2310 — source-bound exact rational RREF/rank/nullspace results.

### PR DESCRIPTION
**Pull request by morluto**
Closes #2318.

## Problem

Jacobian can evaluate or optimize unrestricted `+1/-1` discrepancy and can compute exact rational nullspaces, but neither operation establishes the joint postcondition needed by the cited floating-variable argument: round a rational vector to bits, preserve every disjoint integral row sum exactly, and bound every monitored column error by `4d`.

Leaving the iteration to callers would require each caller to choose retained equations, move to exact endpoints, and reconstruct the global error proof. A mixed-integer solver would return feasibility without the source-derived guarantee.

## Solution

This adds `discrepancy.hard_constraint_round.compute`. Its typed source carries an ordered rational coordinate vector, a named row partition, and named monitored zero-one columns. The result retains that source and returns the binary vector, derived incidence `d`, the `4d` bound, and complete exact row and column ledgers.

The private kernel follows the pinned Erdős construction. At each iteration it retains every active hard row and every monitored column with more than `4d` fractional coordinates, chooses a canonical nonzero direction in the exact nullspace, fixes its primitive sign convention, and takes the first positive cube endpoint. Each step makes at least one coordinate integral. The result validator replays every row sum, column sum, signed error, absolute error, and derived bound.

SymPy 1.14 `DomainMatrix` over `ZZ` is the maintained private nullspace backend. The retained matrices are zero-one integer matrices, and `DomainMatrix.nullspace()` provides an exact fraction-free basis; Jacobian fixes the remaining basis-row, scale, sign, and endpoint conventions for deterministic output. Public rationals and endpoint arithmetic remain canonical `Fraction` values, so backend objects do not leak into the contract.

The public guarantee remains `4d`, matching the pinned proof and this kernel. Doerr proves a sharper general `2d` result, but adopting that theorem would broaden the admitted claim and require a separately justified algorithm and evidence; this PR does not present `4d` as optimal.

The execution envelope is deliberately conservative. It admits at most 512 coordinates, rows, and monitored columns, 32,768 column incidences, 256 digits per input rational, an 8,000,000-unit scan/elimination proxy, 8,192 intermediate rational digits, and a 100,000-digit exact-result budget. Admission derives work from fractional support and incidences and derives height from common denominators plus Hadamard bounds on retained-matrix minors. The kernel checks the admitted direction and height before every endpoint update. These are safe fallback caps, not claims about the largest feasible instance.

The cited source does not publish a materialized allocation fixture in this generic request shape, so the suite does not claim a source-scale reproduction. It instead exercises the defining invariants, an adversarial retained-column case, exhaustive small half-integral systems, exact admission boundaries, authored-result mutations, and coherent relabeling.

Suggested review order:

1. Read `_models.py` for source canonicalization, preflight bounds, and result replay.
2. Read `_operations.py` for the rank count, canonical direction, endpoint step, and termination invariant.
3. Check the adversarial, exhaustive, mutation, and boundary cases in `test_discrepancy_theory.py`.
4. Confirm catalog declaration and owner-local admission in `_tools.py` and `_admission.py`.

## Testing

- `make test-math TESTS=tests/math/discrepancy_theory/test_discrepancy_theory.py` — 26 passed.
- `make test-integration TESTS="tests/integration/catalog/test_builtin_examples.py::test_advertised_invocation_example_executes_successfully[discrepancy.hard_constraint_round.compute] tests/integration/catalog/test_public_operation_contract_conformance.py::test_every_accepted_boundary_mutation_returns_the_declared_result[discrepancy.hard_constraint_round.compute]"` — 2 passed.
- `make check` — Ruff, formatting, and mypy passed; 2,669 bounded-owner tests passed.
- `git diff --check` — passed.
- Deterministic randomized replay — 1,000 exact rational partitioned sources preserved every row and satisfied every monitored-column ledger and bound.
- Boundary measurements — the 74-fractional-coordinate work boundary completed in 0.70 s at 79,628 KiB peak RSS; a retained-column boundary case completed in 0.63 s at 79,712 KiB peak RSS.

- Specialist validation run: the owning discrepancy theory test and the two catalog contract nodes listed above.

## Public contract impact

Adds operation ID `discrepancy.hard_constraint_round.compute`, its request/result schemas, the native typed computation, one executable catalog example, and exact `4d` mathematical semantics. Existing operation contracts are unchanged.

## Public operation admission

- Concrete gap: exact source-bound binary rounding that jointly preserves disjoint integral cardinality constraints and certifies a column-sparse error bound.
- Why existing operations or typed values are insufficient: discrepancy coloring lacks a fractional target and hard rows; rational nullspace returns only one algebraic substep and no global rounding guarantee.
- Stable mathematical result: a retained source, deterministic bit vector, derived `d` and `4d`, and complete exact row/column replay ledgers.
- Semantic domain and admitted execution envelope: rational values in `[0,1]`, an exact named row partition with integral sums, and finite named zero-one columns, under the coordinate, incidence, rational-height, work, intermediate, and result budgets above.
- Controlling work, intermediate, memory, and output quantities: fractional support, coordinate and incidence scans, cumulative exact elimination, common-denominator growth, minor height, and serialized ledger size.
- Exact representation, algorithm regimes, and maintained backend: canonical rationals and bits publicly; one exact floating-variable regime using pinned SymPy 1.14 `DomainMatrix` over `ZZ` privately.
- Remaining fixed caps and their classification: 512-element structural caps and conservative 256/8,192/100,000-digit plus 8,000,000-work safety envelopes; widening them needs stronger analysis or measurements, not another operation.
- Admission decision: `KEEP` in the discrepancy theory owner because this establishes a distinct reusable source-bound rounding postcondition.

## Closure matrix

None. Issue #2318 requests this single admitted operation and leaves no residual public surface.

## Checklist

- [x] `make check` passes
- [x] Explicitly relevant specialist validation is listed above
- [x] Harbor task or verifier changes are not applicable
- [x] Public operation changes include an owner-local admission decision
- [x] The result is exact and source-bound; incomplete, unknown, and unavailable outcomes are not part of this accepted domain
- [x] New bounds include exact boundary tests, adversarial retained-column coverage, and measured boundary cases; the source-scale fixture limitation is stated above
- [x] No shared abstraction was introduced; this operation remains domain-owned

### CODE REVIEW

File: src/jacobian/math/discrepancy_theory/_models.py [hunk 1]
```diff
@@ -2,15 +2,314 @@
...
+                column.coordinates,
+                coordinate_count=coordinate_count,
+                owner="column",
+            )
+            total_incidences += len(column.coordinates)
+            column_sums.append(
+                sum(
+                    (fractions[index] for index in column.coordinates),
+                    Fraction(),
+                )
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/discrepancy_theory/_models.py]: Fixed in `fc8c9c004`. The source validator now completes row/column support checks, rejects cumulative incidence overflow immediately, and checks scan/elimination work plus intermediate rational height before constructing any exact row or column sum. The result-size pass reuses the admitted sums instead of recomputing rows.

The new regression uses 512 high-denominator coordinates and the first full-support column beyond the incidence cap, with exact aggregation replaced by a fail-fast sentinel. It receives the incidence rejection without reaching aggregation.

Validation: the discrepancy owner file passed (27 tests), the focused sentinel passed, the two catalog example/mutation nodes passed, and Ruff, formatting, mypy, and `git diff --check` passed.

[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/discrepancy_theory/_models.py]: Follow-up `83a8cfdce` tightens the regression payload to 256 complementary pairs `(1/D, (D-1)/D)`. The sole hard-row sum is therefore exactly 256 and every non-incidence semantic invariant is valid; under the old ordering, validation would proceed through the full-support column aggregations before rejecting incidence. The focused sentinel and all 27 discrepancy owner tests pass on this corrected fixture.

[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/discrepancy_theory/_models.py]: Verified fixed on this branch (fc8c9c004 + 83a8cfdce): row/column support and partition checks run first, cumulative incidence rejects inside the accumulation loop (_models.py:145-156), scan/elimination-work and denominator-height bounds precede any exact row/column Fraction construction (:158-191). The regression test_over_incidence_rejects_before_exact_aggregation fails loudly if aggregation is reached; an independent probe of the reviewer's shape (512 distinct 256-digit-denominator rationals, over-cap incidences) rejects in ~22 ms with zero exact aggregations executed. Resolving.

**Comment by morluto** [comment]
@codex
